File: src/lib/api/alta.ts
import { invoke } from '@tauri-apps/api/core';

/// 类型定义
export interface AltaQueryResult {
  code: string;
  status: 'forbidden' | 'safe';
  description: string;
  matched_items?: MatchedItem[];
}

export interface MatchedItem {
  code: string;
  description: string;
  level: number;
}

export interface UpdateResult {
  success: boolean;
  items_count: number;
  message: string;
}

export interface DatabaseInfo {
  total_items: number;
  last_update?: string;
  db_size: number;
}

export interface ExcelStats {
  total: number;
  forbidden: number;
  safe: number;
  invalid: number;
  output_path: string;
}

export interface CommandError {
  message: string;
  code: string;
}

/// Alta API 服务
export const altaApi = {
  /**
   * 单个HS编码查询
   * @param hsCode HS编码
   * @param matchLength 匹配位数（4/6/8/undefined为完全匹配）
   */
  async querySingle(
    hsCode: string,
    matchLength?: number
  ): Promise<AltaQueryResult> {
    try {
      return await invoke<AltaQueryResult>('query_hs_code', {
        hsCode,
        matchLength,
      });
    } catch (error) {
      console.error('查询失败:', error);
      throw error;
    }
  },

  /**
   * 更新Alta数据库
   */
  async updateDatabase(): Promise<UpdateResult> {
    try {
      return await invoke<UpdateResult>('update_alta_database');
    } catch (error) {
      console.error('更新数据库失败:', error);
      throw error;
    }
  },

  /**
   * 批量处理Excel文件
   * @param inputPath 输入文件路径
   * @param matchLength 匹配位数
   */
  async batchProcess(
    inputPath: string,
    matchLength?: number
  ): Promise<ExcelStats> {
    try {
      return await invoke<ExcelStats>('batch_process_excel', {
        inputPath,
        matchLength,
      });
    } catch (error) {
      console.error('批量处理失败:', error);
      throw error;
    }
  },

  /**
   * 获取数据库信息
   */
  async getDatabaseInfo(): Promise<DatabaseInfo> {
    try {
      return await invoke<DatabaseInfo>('get_database_info');
    } catch (error) {
      console.error('获取数据库信息失败:', error);
      throw error;
    }
  },

  /**
   * 下载Excel模板
   */
  async downloadTemplate(outputPath: string): Promise<void> {
    try {
      await invoke('download_template', { outputPath });
    } catch (error) {
      console.error('下载模板失败:', error);
      throw error;
    }
  },

  /**
   * 测试数据库连接
   */
  async testDatabaseConnection(): Promise<boolean> {
    try {
      return await invoke<boolean>('test_database_connection');
    } catch (error) {
      console.error('测试数据库连接失败:', error);
      return false;
    }
  },

  /**
   * 测试Alta网站连接
   */
  async testAltaConnection(): Promise<boolean> {
    try {
      return await invoke<boolean>('test_alta_connection');
    } catch (error) {
      console.error('测试Alta连接失败:', error);
      return false;
    }
  },
};

File: src-tauri/src/models/alta.rs
use serde::{Deserialize, Serialize};

/// 禁运商品模型
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForbiddenItem {
    pub id: Option<i64>,
    pub hs_code: String,
    pub hs_code_4: String,
    pub hs_code_6: String,
    pub hs_code_8: String,
    pub description: String,
    pub additional_info: String,
    pub source_url: String,
    pub created_at: Option<String>,

    // === 新增字段 (v2) ===
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raw_text: Option<String>,          // 原始 HS 编码文本

    #[serde(skip_serializing_if = "Option::is_none")]
    pub has_exceptions: Option<bool>,      // 是否包含例外
}

/// HS 编码条目（用于解析结果）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HsCodeEntry {
    pub code: String,
    pub code_4: String,
    pub code_6: String,
    pub code_8: String,
    pub is_exception: bool,
    pub parent_raw: String,
}

/// 匹配结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MatchResult {
    pub is_forbidden: bool,
    pub matched_codes: Vec<String>,
    pub descriptions: Vec<String>,
    pub match_type: String,
    pub raw_texts: Vec<Option<String>>,      // 原始文本列表
    pub has_exceptions: Vec<bool>,            // 是否包含例外列表
}

/// 查询请求
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryRequest {
    pub hs_code: String,
    pub match_length: Option<u8>, // 4, 6, 8, or None for exact
}

/// 更新结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateResult {
    pub success: bool,
    pub items_count: usize,
    pub message: String,
}

/// 数据库信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseInfo {
    pub total_items: i64,
    pub last_update: Option<String>,
    pub db_size: u64,
}

/// Excel 处理统计
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExcelStats {
    pub total: usize,
    pub forbidden: usize,
    pub safe: usize,
    pub invalid: usize,
    pub output_path: String,
}

/// 匹配的商品项（用于前端显示）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MatchedItem {
    pub code: String,
    pub description: String,
    pub level: u8,
    pub raw_text: Option<String>,         // 原始 HS 编码文本
    pub has_exceptions: Option<bool>,     // 是否包含例外
}

/// Alta 查询结果（前端接口）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AltaQueryResult {
    pub code: String,
    pub status: String, // "forbidden" or "safe"
    pub description: String,
    pub matched_items: Option<Vec<MatchedItem>>,
}

// ============================================================================
// ForbiddenItem 辅助方法
// ============================================================================

impl ForbiddenItem {
    /// 创建 v1 格式的条目（向后兼容）
    pub fn new_v1(
        hs_code: String,
        description: String,
        additional_info: String,
    ) -> Self {
        let hs_code_4 = Self::prefix(&hs_code, 4);
        let hs_code_6 = Self::prefix(&hs_code, 6);
        let hs_code_8 = Self::prefix(&hs_code, 8);

        Self {
            id: None,
            hs_code,
            hs_code_4,
            hs_code_6,
            hs_code_8,
            description,
            additional_info,
            source_url: String::new(),
            created_at: None,
            raw_text: None,
            has_exceptions: None,
        }
    }

    /// 创建 v2 格式的条目（含完整信息）
    pub fn new_v2(
        code: HsCodeEntry,
        raw_text: String,
        has_exceptions: bool,
        description: String,
        additional_info: String,
        source_url: String,
    ) -> Self {
        let enhanced_description = if has_exceptions {
            format!("{} [含例外]", description)
        } else {
            description
        };

        Self {
            id: None,
            hs_code: code.code.clone(),
            hs_code_4: code.code_4,
            hs_code_6: code.code_6,
            hs_code_8: code.code_8,
            description: enhanced_description,
            additional_info,
            source_url,
            created_at: None,
            raw_text: Some(raw_text),
            has_exceptions: Some(has_exceptions),
        }
    }

    fn prefix(code: &str, len: usize) -> String {
        if code.len() >= len {
            code[..len].to_string()
        } else {
            code.to_string()
        }
    }
}
